File: core/src/main/scala/com/github/propi/rdfrules/ruleset/ops/Sortable.scala
package com.github.propi.rdfrules.ruleset.ops

import com.github.propi.rdfrules.data.ops.Transformable
import com.github.propi.rdfrules.ruleset.ops.Sortable.Sorted
import com.github.propi.rdfrules.utils.ForEach

import scala.collection.mutable

/**
  * Created by Vaclav Zeman on 17. 4. 2018.
  */
trait Sortable[T, Coll] extends Transformable[T, Coll] {

  protected implicit val ordering: Ordering[T]

  def sorted: Coll = transform(new Sorted(coll))

  def sortBy[A](f: T => A)(implicit ord: Ordering[A]): Coll = transform(new Sorted(coll)(Ordering.by[T, A](f)))

  def reversed: Coll = transform((f: T => Unit) => coll.toIndexedSeq.reverseIterator.foreach(f))

}

object Sortable {

  private class Sorted[T](col: ForEach[T])(implicit ord: Ordering[T]) extends ForEach[T] {
    private lazy val sorted = {
      val queue = mutable.PriorityQueue.empty[T]
      col.foreach(queue.enqueue(_))
      queue.dequeueAll.asInstanceOf[IndexedSeq[T]]
    }

    def foreach(f: T => Unit): Unit = sorted.foreach(f)

    override def knownSize: Int = sorted.knownSize
  }

}

File: java/src/main/java/com/github/propi/rdfrules/java/index/Sortable.java
package com.github.propi.rdfrules.java.index;

import com.github.propi.rdfrules.java.OrderingWrapper;
import com.github.propi.rdfrules.java.data.Transformable;

import java.util.Comparator;
import java.util.function.Function;

/**
 * Created by Vaclav Zeman on 13. 5. 2018.
 */
public interface Sortable<ST, JT, SColl, JColl> extends Transformable<ST, JT, SColl, JColl> {

    com.github.propi.rdfrules.ruleset.ops.Sortable<ST, SColl> asScala();

    default JColl sorted() {
        return asJava(asScala().sorted());
    }

    default <T> JColl sortBy(Function<JT, T> f, Comparator<T> comparator) {
        return asJava(asScala().sortBy(x -> f.apply(asJavaItem(x)), new OrderingWrapper<>(comparator)));
    }

}